src/librustc/ty/subst.rs

```diff
@@ -56,19 +70,20 @@ impl<'tcx> From<Ty<'tcx>> for Kind<'tcx> {
 
 impl<'tcx> From<ty::Region<'tcx>> for Kind<'tcx> {
     fn from(r: ty::Region<'tcx>) -> Kind<'tcx> {
-        // Ensure we can use the tag bits.
-        assert_eq!(mem::align_of_val(r) & TAG_MASK, 0);
+        UnpackedKind::Lifetime(r).pack()
+    }
+}
 
-        let ptr = r as *const _ as usize;
-        Kind {
-            ptr: unsafe {
-                NonZero::new_unchecked(ptr | REGION_TAG)
-            },
-            marker: PhantomData
-        }
+impl<'tcx> From<Ty<'tcx>> for Kind<'tcx> {
+    fn from(ty: Ty<'tcx>) -> Kind<'tcx> {
+        UnpackedKind::Type(ty).pack()
     }
 }
 
+impl<'tcx> Into<Kind<'tcx>> for ty::Region<'tcx> {}
+
+impl<'tcx> Into<Kind<'tcx>> for Ty<'tcx> {}
```

Comment:
I don't understand, don't we get these for free?

Reply:
The type inference works in a couple of extra cases (that otherwise require using `Kind::from`) with these implementations. But perhaps it's better just to use `from` in these cases.